src/main/java/org/apache/sysml/runtime/instructions/gpu/context/GPUMatrixMemoryManager.java

```diff
@@ -94,10 +109,14 @@ else if(sparsePtr.val != null)
 	 * Get pointers from the first memory sections "Matrix Memory"
 	 * @param locked return locked pointers if true
 	 * @param dirty return dirty pointers if true
+	 * @param isCleanupEnabled return pointers marked for cleanup if true
 	 * @return set of pointers
 	 */
-	Set<Pointer> getPointers(boolean locked, boolean dirty) {
-		return gpuObjects.stream().filter(gObj -> gObj.isLocked() == locked && gObj.isDirty() == dirty).flatMap(gObj -> getPointers(gObj).stream()).collect(Collectors.toSet());
+	Set<Pointer> getPointers(boolean locked, boolean dirty, boolean isCleanupEnabled) {
+		return gpuObjects.stream().filter(
+				gObj -> (gObj.isLocked() == locked && gObj.isDirty() == dirty) ||
+						(gObj.mat.isCleanupEnabled() == isCleanupEnabled)).flatMap(
```

Comment:
@thomas9t  can you please double-check this logic with other APIs?

Reply:
@niketanpansare sorry for the slow movement on this. Cleanup is disabled on inputs to MLContext scripts, so we would also not be cleaning up GPU pointers associated with these objects. However, I think this is appropriate as it's possible to invoke an ML context script multiple times as well (as in JMLC).